npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb

```diff
@@ -256,7 +257,7 @@ def requirement_for(requirement)
         return requirement unless git_url?(requirement)
 
         details = requirement.match(GIT_URL_REGEX).named_captures
-        details["semver"]
+        details["semver"] || (requirement.start_with?("github:") ? requirement : nil)
```

Comment:
Do we need to change this? Wondering if we should leave this as is, being `nil` if it doesn't match semver. Doesn't look like we actually rely on this change anywhere and might cause issues upstream issues when creating pull requests.

Reply:
I think this steer is a good idea but unsure of just changing it here for requirements that specifically match `github:`. Might be a good idea to figure out if we could always insert the requirement here but feels out of scope for this change.

Reply:
> Do we need to change this?

Yes. I believe this this is the root cause of the issue. A `nil` constraint doesn't provide `npm` with enough information. More info here https://github.com/dependabot/dependabot-core/issues/3813#issuecomment-853458082.

I would prefer to change this line to:

```ruby
  details["semver"] || requirement
```

This would ensure that we always pass the original constraint down to npm to complete an install using the constraint provided in the manifest. However, I'm unsure of the other consequences of this change. Because of that I chose to be defensive and try to restrict this to only [GitHub repository style URLs](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#repository) for the time being. Without this change the defect is consistently reproduced.

Reply:
Ah not sure this gets used in the npm update, the requirement is extracted from the `package.json` file here https://github.com/dependabot/dependabot-core/blob/0b514b24ff8a201e379157a22a50107f7d86af27/npm_and_yarn/helpers/lib/npm6/updater.js#L52

Confusingly we also do this here https://github.com/dependabot/dependabot-core/blob/main/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb#L252-L257 🤔

I think this might need to broader refactors to clean up, I definitely think it would be a lot better to pass this from the file parser instead of re-extracting it in the lockfile updater and native helper.

Reply:
I am looking for more context around how the requirement is used in the downstream processing. Please do tell me what you think and if you have any alternative ideas that I can/should pursue.

Reply:
Does the dry-run update work if you revert this change? If so would be up for shipping this, I have a feeling this change doesn't have an impact on the update so might just cause issues in the pr creator where we use the requirement to build the PR title/body

Reply:
> Without this change the defect is consistently reproduced

I tested this dry-run without this change and seems to work but might be missing something else?

```
bin/dry-run.rb npm_and_yarn Hypixel-Translators/hypixel-translators-bot --dep=discord.js --cache=files --commit=e2bfc6768492d32f22bdc6a732237315b6b22dd2
```

Reply:
> Ah not sure this gets used in the npm update, the requirement is extracted from the `package.json` file here
> 
> https://github.com/dependabot/dependabot-core/blob/0b514b24ff8a201e379157a22a50107f7d86af27/npm_and_yarn/helpers/lib/npm6/updater.js#L52
> 
> Confusingly we also do this here https://github.com/dependabot/dependabot-core/blob/main/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb#L252-L257 🤔
> 
> I think this might need to broader refactors to clean up, I definitely think it would be a lot better to pass this from the file parser instead of re-extracting it in the lockfile updater and native helper.

You're right. We don't need this. I just re-tested with and without and both appear to leave the `from` field in the desired state. I'll remove this change. Thank you for spotting this @feelepxyz. 🙇